### PR TITLE
Add ability to watch secondary resources

### PR DESCRIPTION
**Pull request by fabianlindfors**
### All Submissions:

* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change

This PR implements a custom `handler.EventHandler` called `ResourceWatcher`. This will allow us to watch for changes to specific secondary resources outside of the main MDB CRD. This PR is part of the work to implement support for TLS certificate rotations. The idea is to have one `ResourceWatcher` on the `Reconciler` for each resource type we want to watch. If we for example have a `secretWatcher` on our `Reconciler` we can register it like:

```
// Watch for changes to primary resource MongoDB
err = c.Watch(&source.Kind{Type: &mdbv1.MongoDB{}}, &handler.EnqueueRequestForObject{}, predicates.OnlyOnSpecChange())
if err != nil {
	return err
}

// Watch for changes to secondary Secrets
err = c.Watch(&source.Kind{Type: &corev1.Secret{}}, r.secretWatcher)
if err != nil {
	return err
}
```

Resource to watch can then be added from the reconciliation loop using: `r.secretWatcher.Watch(objectToWatch, dependentObject)`.

### CODE REVIEW

File: pkg/controller/watch/watch.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+// New will create a new ResourceWatcher with no watched objects.
+func New() ResourceWatcher {
+	return ResourceWatcher{
+		watched: make(map[types.NamespacedName][]types.NamespacedName),
+	}
+}
+
+// Add will add a new object to watch.
+func (w ResourceWatcher) Add(watchedName, dependentName types.NamespacedName) {
```
Comment (chatton) [inline pkg/controller/watch/watch.go:26]: [nit] The name `Add` is a bit misleading, I think maybe something like `EnsureAdded` or `EnsureWatched` would be more accurate, what do you think?

File: pkg/controller/watch/watch.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	if !hasExisting {
+		existing = []types.NamespacedName{}
+	}
+
+	// Check if resource is already being watched.
+	for _, existingName := range existing {
+		if dependentName == existingName {
+			return
+		}
+	}
```
Comment (chatton) [inline pkg/controller/watch/watch.go:37]: we cloud maybe add a  `contains.NamespacedName(nsNames, nsName)` for this.

Comment (fabianlindfors) [reply pkg/controller/watch/watch.go:37]: True, added!

File: pkg/controller/watch/watch.go (hunk 1 above)
Comment (fabianlindfors) [reply pkg/controller/watch/watch.go:27]: I agree that `Add` is a bit misleading but I'm not sure including `Ensure` adds much. How about `Watch`? Feels natural that you would `Watch` with a `Watcher`!

Comment (chatton) [reply pkg/controller/watch/watch.go:27]: I think I'm okay with Watch!

File: pkg/util/contains/contains.go [hunk 3]
```diff
@@ -19,3 +22,12 @@ func AuthMode(slice []mdbv1.AuthMode, s mdbv1.AuthMode) bool {
...
 }
+
+func NamespacedName(nsNames []types.NamespacedName, nsName types.NamespacedName) bool {
+	for _, elem := range nsNames {
+		if elem == nsName {
+			return true
+		}
+	}
+	return false
+}
```
Comment (chatton) [inline pkg/util/contains/contains.go:33]: 👍

File: pkg/controller/watch/watch.go (hunk 1 above)
Comment (fabianlindfors) [reply pkg/controller/watch/watch.go]: Renamed!

**Pull request merged**